<a href="https://colab.research.google.com/github/hjhan1201/AIFFEL_Online_Quest/blob/main/DeepLearning_from_Scratch_Ch2~3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
#퍼셉트론 구현하기
def AND(x1, x2) :
  w1, w2, theta = 0.5,0.5,0.7
  tmp = x1*w1 + x2*w2
  if tmp <= theta:
    return 0
  elif tmp >theta:
    return 1

AND(1,1)

#넘파이와 b(bias,편향) 적용
import numpy as np
x = np.array([0,1])
w = np.array([0.5,0.5])
b = -0.7
w*x
np.sum(w*x)
np.sum(w*x)+b

#가중치와 편향을 도입한 AND게이트
def AND(x1, x2) :
  x = np.array([x1,x2])
  w = np.array([0.5,0.5])
  b = -0.7
  tmp = np.sum(w*x)+b
  if tmp <= 0:
    return 0
  elif tmp > 0:
    return 1

AND(1,1)

#NAND 게이트
def NAND(x1, x2) :
  x = np.array([x1,x2])
  w = np.array([-0.5,-0.5])
  b = 0.7
  tmp = np.sum(w*x)+b
  if tmp <= 0:
    return 0
  elif tmp > 0:
    return 1

#OR 게이트
def OR(x1, x2) :
  x = np.array([x1,x2])
  w = np.array([0.5,0.5])
  b = -0.2
  tmp = np.sum(w*x)+b
  if tmp <= 0:
    return 0
  elif tmp > 0:
    return 1

OR(0,1)

#XOR(단층 퍼셉트론으로 구현 불가)
def XOR(x1,x2):
  s1 = NAND(x1, x2)
  s2 = OR(x1, x2)
  y = AND(s1, s2)
  return y


XOR(0,0)
XOR(0,1)
XOR(1,0)
XOR(1,1)

#Ch3
#계단함수 구현하기
def setp_function(x):
  if x>0:
    return 1
  elif x<=0:
    return 0

#넘파이 배열도 인수로 넣고 싶다
def step_function(x):
  y=x>0
  return y.astype(np.int)

x=np.array([-1.0,1.0,2.0])
x
y=x>0
y = y.astype(int)
y

#계단함수 그려보기
import matplotlib.pylab as plt

def step_function(x):
  return np.array(x>0,dtype=np.int)
x = np.arange(-5.0,5.0,0.1)
y = step_function(x)
# plt.plot(x, y, ls='--')
# plt.ylim(-0.1,1.1) #y축의 범위 지정
# plt.show

#시그모이드 함수 구현하기
def sigmoid(x):
  return 1/(1+np.exp(-x))
x = np.array([-1.0,1.0,2.0])
sigmoid(x)

#넘파이 브로드캐스트
t = np.array([1.0,2.0,3.0])
1.0/t

#시그모이드함수 그려보기
x = np.arange(-5.0,5.0,0.1)
y = sigmoid(x)
# plt.plot(x,y)
# plt.ylim(-0.1,1.1) #y축의 범위 지정
# plt.show

#ReLU 함수
def relu(x):
  return np.maximum(0,x)

#행렬의 곱
A = np.array([[1,2],[3,4]])
B = np.array([[5,6],[7,8]])
np.dot(A,B)

A = np.array([[1,2],[3,4],[5,6]])
B = np.array([7,8])
np.dot(A,B)

X = np.array([1,2])
W = np.array([[1,3,5],[2,4,6]])
Y = np.dot(X,W)
Y

#신경망 행렬곱
X = np.array([1.0,0.5])
W1 = np.array([[0.1,0.3,0.5],[0.2,0.4,0.6]])
B1 = np.array([0.1,0.2,0.3])

A1 = np.dot(X,W1)+B1
Z1 = sigmoid(A1)
Z1

W2 = np.array([[0.1,0.4],[0.2,0.5],[0.3,0.6]])
B2 = np.array([0.1,0.2])
A2 = np.dot(Z1,W2)+B2
Z2 = sigmoid(A2)
Z2

def identity_function(x): #굳이 필요 없음
  return x

W3 = np.array([[0.1,0.3],[0.2,0.4]])
B3 = np.array([0.1,0.2])
A3 = np.dot(Z2,W3)+B3
Z3 = sigmoid(A3)
Y = identity_function(A3)
# print(Y)
# print(Z3)

#신경망 구현 정리
def init_network():
    network = {}
    network['W1'] = np.array([[0.1, 0.3, 0.5], [0.2, 0.4, 0.6]])
    network['b1'] = np.array([0.1, 0.2, 0.3])
    network['W2'] = np.array([[0.1, 0.4], [0.2, 0.5], [0.3, 0.6]])
    network['b2'] = np.array([0.1, 0.2])
    network['W3'] = np.array([[0.1, 0.3], [0.2, 0.4]])
    network['b3'] = np.array([0.1, 0.2])

    return network

def forward(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = identity_function(a3)

    return y

network = init_network()
x = np.array([1.0, 0.5])
y = forward(network, x)
# print(y) # [0.31682708 0.69627909]



<ipython-input-19-bbb7787ee6eb>:94: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  return np.array(x>0,dtype=np.int)


In [20]:
#소프트맥스 함수 구현
a = np.array([0.3,2.9,4.0])

exp_a = np.exp(a)
print(exp_a)
sum_exp_a = np.sum(exp_a)
print(sum_exp_a)

y = exp_a / sum_exp_a
print(y)

def softmax(a) :
  exp_a = np.exp(a)
  sum_exp_a = np.sum(exp_a)
  y = exp_a / sum_exp_a
  return y



[ 1.34985881 18.17414537 54.59815003]
74.1221542101633
[0.01821127 0.24519181 0.73659691]


In [21]:
a = np.array([1010, 1000, 990])
np.exp(a) / np.sum(np.exp(a)) # 소프트맥스 함수의 계산
# array([ nan,  nan,  nan])   # 제대로 계산되지 않는다.

c = np.max(a)
a - c
np.exp(a - c) / np.sum(np.exp(a - c))


def softmax(a):
    c = np.max(a)
    exp_a = np.exp(a-c) # 오버플로 대책
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a

    return y

a = np.array([0.3, 2.9, 4.0])
y = softmax(a)
print(y)
np.sum(y)

[0.01821127 0.24519181 0.73659691]


<ipython-input-21-493c4fba24de>:2: RuntimeWarning: overflow encountered in exp
  np.exp(a) / np.sum(np.exp(a)) # 소프트맥스 함수의 계산
<ipython-input-21-493c4fba24de>:2: RuntimeWarning: invalid value encountered in divide
  np.exp(a) / np.sum(np.exp(a)) # 소프트맥스 함수의 계산


1.0

소프트맥스 함수 출력의 총합은 1. 출력을 '확률'로 해석할 수 있음.

y[0]의 확률 0.018(1.8%), y[1]의 확률 0.245(24.5%), y[2]의 확률 0.737(73.7%)로 해석가능

"2번째 원소의 확률이 가장 높으니, 답은 2번 클래스다"

소프트맥스 함수를 이용함으로써 문제를 확률적(통계적)으로 대응할 수 있게 됨

소프트맥스를 적용해도 각 원소의 대소 관계는 변하지 않음

이유: 단조 증가함수(a<=b일 때 f(a)<=f(b)가 성립하는 함수)이기 때문

추론 단계에서는 출력층의 소프트맥스 함수를 생략하는 것이 일반적

학습 시킬 때에는 출력층에서 소프트맥스 함수를 사용

In [39]:
import sys, os
sys.path.append(os.pardir) # 부모 디렉터리의 파일을 가져올 수 있도록 설정
sys.path.append("/content")
from mnist import load_mnist

import pickle

def get_data():
    (x_train, t_train), (x_test, t_test) = load_mnist(normalize=True, flatten=True, one_hot_label = False)
    return x_test, t_test

def init_network():
    with open("sample_weight.pkl", 'rb') as f:
        network = pickle.load(f)

    return network

def predict(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = softmax(a3)

    return y

x, t = get_data()
network = init_network()

accuracy_cnt = 0
for i in range(len(x)):
    y = predict(network, x[i])
    p = np.argmax(y) # 확률이 가장 높은 원소의 인덱스를 얻음
    if p == t[i]:
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

FileNotFoundError: ignored